In [2]:
import pandas as pd

In [3]:
df = pd.read_csv("pedidos_loja.csv")

In [4]:
df.head()

,id_pedido,data_pedido,id_cliente,canal,cidade,estado,categoria,produto,quantidade,preco_unitario,frete,status,data_entrega
0,PED-01169,2024-10-17,CLI-0394,site,Salvador,BA,esporte,Corda de Pular Speed,1,67.34,"R$ 28,06",entregue,2024-10-28
1,PED-01204,2024-08-02,CLI-0315,app,Brasília,DF,Casa & Decoração,"Tapete Sala 1,5m",4,214.43,"R$ 0,00",entregue,2024-08-16
2,PED-01464,2023-11-06,CLI-0223,app,Curitiba,PR,Moda,Jaqueta Jeans,4,188.97,"R$ 0,00",entregue,2023-11-16
3,PED-04480,2024-12-19,CLI-1293,app,Curitiba,PR,Beleza,Secador de Cabelo 2000W,1,192.99,"R$ 20,40",entregue,2025-01-02
4,PED-03605,2024-12-04,CLI-0958,site,Porto Alegre,RS,Beleza,Secador de Cabelo 2000W,2,336.47,"R$ 0,00",entregue,2024-12-16


In [5]:
df.shape

(5118, 13)

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5118 entries, 0 to 5117
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id_pedido       5118 non-null   str    
 1   data_pedido     5118 non-null   str    
 2   id_cliente      5118 non-null   str    
 3   canal           5118 non-null   str    
 4   cidade          5087 non-null   str    
 5   estado          4955 non-null   str    
 6   categoria       5062 non-null   str    
 7   produto         5118 non-null   str    
 8   quantidade      5118 non-null   int64  
 9   preco_unitario  5047 non-null   float64
 10  frete           5118 non-null   str    
 11  status          5118 non-null   str    
 12  data_entrega    4249 non-null   str    
dtypes: float64(1), int64(1), str(11)
memory usage: 519.9 KB


In [7]:
df.describe()

,quantidade,preco_unitario
count,5118.000000,5047.000000
mean,1.961899,213.873662
std,1.098222,156.525215
min,-2.000000,-760.500000
25%,1.000000,100.310000
50%,2.000000,173.600000
75%,3.000000,288.355000
max,4.000000,898.360000


In [8]:
df.isna().sum()

id_pedido           0
data_pedido         0
id_cliente          0
canal               0
cidade             31
estado            163
categoria          56
produto             0
quantidade          0
preco_unitario     71
frete               0
status              0
data_entrega      869
dtype: int64

In [9]:
df.duplicated().sum()

np.int64(118)

In [10]:
df["status"].value_counts()

status
entregue         3442
em transporte     421
cancelado         365
devolvido         282
Entregue          247
ENTREGUE          240
CANCELADO          29
Em transporte      22
Devolvido          22
EM TRANSPORTE      17
DEVOLVIDO          16
Cancelado          15
Name: count, dtype: int64

In [11]:
df["cidade"].nunique()

52

In [12]:
sorted(df["cidade"].dropna().unique())

['  Belo Horizonte ',
 '  Brasília ',
 '  Campinas ',
 '  Curitiba ',
 '  Fortaleza ',
 '  Goiânia ',
 '  Porto Alegre ',
 '  Recife ',
 '  Ribeirão Preto ',
 '  Rio de Janeiro ',
 '  Salvador ',
 '  São Paulo ',
 'BELO HORIZONTE',
 'BRASÍLIA',
 'Belo Horizonte',
 'Brasilia',
 'Brasília',
 'CAMPINAS',
 'CURITIBA',
 'Campinas',
 'Curitiba',
 'FORTALEZA',
 'Fortaleza',
 'GOIÂNIA',
 'Goiania',
 'Goiânia',
 'PORTO ALEGRE',
 'Porto Alegre',
 'RECIFE',
 'RIBEIRÃO PRETO',
 'RIO DE JANEIRO',
 'Recife',
 'Ribeirao Preto',
 'Ribeirão Preto',
 'Rio de Janeiro',
 'SALVADOR',
 'Salvador',
 'Sao Paulo',
 'SÃO PAULO',
 'São Paulo',
 'belo horizonte',
 'brasília',
 'campinas',
 'curitiba',
 'fortaleza',
 'goiânia',
 'porto alegre',
 'recife',
 'ribeirão preto',
 'rio de janeiro',
 'salvador',
 'são paulo']

## Diagnóstico inicial, 7 perguntas respondidas:
1) Quantas linhas e colunas tem a base?
5118, 13

2) preco_unitario e frete são do mesmo tipo de dado? Deveriam ser?
Não são do mesmo tipo de dado. preco é float64 e frete str. Float64 = número decimal = 74.43. Str = texto puro, pode ter número. Frete deveria ser Float64, pois entregou valores decimais com virgulas em forma de texto, não pode ser lido pelo pandas.

3) Por que describe() só resumiu 2 das 13 colunas?
Pq descibre só calcula colunas numéricas

4) A loja atende 12 cidades reais. Por que nunique() da coluna cidade deu 52?
Porque as cidades estão sem padronização de texto: Salvador x SALVADOR. O comando "sorted(df["cidade"].dropna().unique())" mostrou a diferença da grafia para as mesmas cidades.


5) Por que value_counts() do status mostrou mais de 4 categorias, se só existem 4 status possíveis?
Bom ela pede 4 categorias, mas listou algumas com o mesmo nome porém grafia diferente, em transporte x EM TRANSPORTE por exemplo. 

6) Existe algum valor de quantidade que não faz sentido pra um pedido?
No describe tem um pedido com quantidade negativa - 2

7) O duplicate().sum() deu 118, o que isso quer dizer na base?
Mostra o número exato de cópias exatas. 


In [13]:
df["frete"] = df["frete"].str.replace("R$ ", "", regex=False)
df["frete"] = df["frete"].str.replace(",", ".", regex=False)
df["frete"] = df["frete"].astype(float)

In [14]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5118 entries, 0 to 5117
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id_pedido       5118 non-null   str    
 1   data_pedido     5118 non-null   str    
 2   id_cliente      5118 non-null   str    
 3   canal           5118 non-null   str    
 4   cidade          5087 non-null   str    
 5   estado          4955 non-null   str    
 6   categoria       5062 non-null   str    
 7   produto         5118 non-null   str    
 8   quantidade      5118 non-null   int64  
 9   preco_unitario  5047 non-null   float64
 10  frete           5118 non-null   float64
 11  status          5118 non-null   str    
 12  data_entrega    4249 non-null   str    
dtypes: float64(2), int64(1), str(10)
memory usage: 519.9 KB


In [15]:
df = df.drop_duplicates()

In [16]:
df.shape

(5000, 13)

In [17]:
df.duplicated().sum()

np.int64(0)

In [18]:
df["cidade"] = df["cidade"].str.strip()
df["cidade"] = df["cidade"].str.title()

In [19]:
df["cidade"].nunique()

16

In [20]:
sorted(df["cidade"].dropna().unique())

['Belo Horizonte',
 'Brasilia',
 'Brasília',
 'Campinas',
 'Curitiba',
 'Fortaleza',
 'Goiania',
 'Goiânia',
 'Porto Alegre',
 'Recife',
 'Ribeirao Preto',
 'Ribeirão Preto',
 'Rio De Janeiro',
 'Salvador',
 'Sao Paulo',
 'São Paulo']

In [21]:
import unicodedata
import pandas as pd

def remove_acentos(texto):
    if pd.isna(texto):
        return texto
    return unicodedata.normalize("NFKD", texto).encode("ascii", "ignore").decode("ascii")

df["cidade"] = df["cidade"].apply(remove_acentos)

In [22]:
df["cidade"].nunique()

12

In [23]:
sorted(df["cidade"].dropna().unique())

['Belo Horizonte',
 'Brasilia',
 'Campinas',
 'Curitiba',
 'Fortaleza',
 'Goiania',
 'Porto Alegre',
 'Recife',
 'Ribeirao Preto',
 'Rio De Janeiro',
 'Salvador',
 'Sao Paulo']

In [24]:
sorted(df["categoria"].dropna().unique())

['BELEZA',
 'Beleza',
 'CASA & DECORAÇÃO',
 'Casa & Decoração',
 'ELETRÔNICOS',
 'ESPORTE',
 'Eletrônicos',
 'Esporte',
 'MODA',
 'Moda',
 'beleza',
 'casa & decoracao',
 'eletronicos',
 'esporte',
 'moda']

In [25]:
sorted(df["status"].dropna().unique())


['CANCELADO',
 'Cancelado',
 'DEVOLVIDO',
 'Devolvido',
 'EM TRANSPORTE',
 'ENTREGUE',
 'Em transporte',
 'Entregue',
 'cancelado',
 'devolvido',
 'em transporte',
 'entregue']

In [26]:
df["status"] = df["status"].str.strip()
df["status"] = df["status"].str.title()

In [27]:
df["categoria"] = df["categoria"].str.strip()
df["categoria"] = df["categoria"].str.title()

In [28]:
df["categoria"] = df["categoria"].apply(remove_acentos)

In [29]:
df["status"].nunique()

4

In [30]:
df["categoria"].nunique()

5

In [31]:
df[df["quantidade"] <= 0]

,id_pedido,data_pedido,id_cliente,canal,cidade,estado,categoria,produto,quantidade,preco_unitario,frete,status,data_entrega
59,PED-00980,2023-07-10,CLI-0254,site,Rio De Janeiro,RJ,Casa & Decoracao,Luminária de Mesa LED,0,127.24,17.22,Cancelado,NaN
105,PED-03945,2023-07-01,CLI-0777,marketplace,Sao Paulo,SP,Esporte,Tapete de Yoga 6mm,0,188.14,16.23,Entregue,2023-07-05
170,PED-00905,2024-12-14,CLI-1344,marketplace,Sao Paulo,SP,Esporte,Garrafa Térmica 1L,-2,71.91,14.89,Entregue,2024-12-28
215,PED-02507,2024-12-06,CLI-0638,site,Sao Paulo,SP,Moda,Tênis Casual Masculino,-1,341.01,0.00,Entregue,2024-12-09
519,PED-00930,2024-08-20,CLI-1348,app,Curitiba,PR,Casa & Decoracao,Jogo de Toalhas 4 peças,0,185.98,0.00,Entregue,2024-09-03
933,PED-01835,2023-08-08,CLI-0533,app,Rio De Janeiro,RJ,Casa & Decoracao,Luminária de Mesa LED,0,198.74,21.52,Entregue,2023-08-17
1038,PED-00113,2023-10-29,CLI-0607,app,Ribeirao Preto,SP,Beleza,Kit Skincare Facial,-1,-260.59,0.00,Entregue,2023-11-04
1059,PED-02896,2023-04-08,CLI-0281,app,Rio De Janeiro,RJ,Eletronicos,Webcam Full HD,-2,324.57,0.00,Entregue,2023-04-11
1129,PED-04077,2024-06-19,CLI-1262,app,Sao Paulo,SP,Esporte,Tapete de Yoga 6mm,-1,171.73,19.57,Cancelado,NaN
1163,PED-00537,2024-03-24,CLI-0428,site,Brasilia,DF,Beleza,Kit Skincare Facial,0,219.78,0.00,Entregue,2024-04-09


In [32]:
quantidade_invalida = df[df["quantidade"] <= 0]
quantidade_invalida.to_csv("quantidade_invalida_revisar.csv", index=False)

In [33]:
df = df[df["quantidade"] > 0]

In [34]:
df.shape

(4955, 13)

In [35]:
df["quantidade"].min()

np.int64(1)

In [36]:
df.isna().sum()

id_pedido           0
data_pedido         0
id_cliente          0
canal               0
cidade             30
estado            160
categoria          55
produto             0
quantidade          0
preco_unitario     70
frete               0
status              0
data_entrega      851
dtype: int64

In [37]:
df[df["data_entrega"].isna()]["status"].value_counts()

status
Em Transporte    452
Cancelado        399
Name: count, dtype: int64

In [38]:
df[df["cidade"].isna()]["estado"].isna().sum()

np.int64(0)

In [39]:
df.groupby("estado")["cidade"].nunique()

estado
BA    1
CE    1
DF    1
GO    1
MG    1
PE    1
PR    1
RJ    1
RS    1
SP    3
Name: cidade, dtype: int64

In [40]:
df[df["cidade"].isna()]["estado"].value_counts()

estado
SP    14
PR     9
CE     2
RJ     2
RS     1
MG     1
PE     1
Name: count, dtype: int64

In [41]:
df.groupby("estado")["cidade"].unique()

estado
BA                                    [Salvador]
CE                              [Fortaleza, nan]
DF                                    [Brasilia]
GO                                     [Goiania]
MG                         [Belo Horizonte, nan]
PE                                 [Recife, nan]
PR                               [Curitiba, nan]
RJ                         [Rio De Janeiro, nan]
RS                           [Porto Alegre, nan]
SP    [Sao Paulo, Campinas, Ribeirao Preto, nan]
Name: cidade, dtype: object

In [42]:
mapa_estado_cidade = {
    "BA": "Salvador",
    "CE": "Fortaleza",
    "DF": "Brasilia",
    "GO": "Goiania",
    "MG": "Belo Horizonte",
    "PE": "Recife",
    "PR": "Curitiba",
    "RJ": "Rio De Janeiro",
    "RS": "Porto Alegre",
}

In [43]:
cidade_recuperada = df["estado"].map(mapa_estado_cidade)
df["cidade"] = df["cidade"].fillna(cidade_recuperada)

In [44]:
df["cidade"].isna().sum()

np.int64(14)